pkg/reconciler/workload/namespace/namespace_reconcile.go

```diff
@@ -244,79 +239,67 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName logica
 		ns.Labels = map[string]string{}
 	}
 
-	ns, err = c.ensureScheduled(ctx, ns)
+	ns, rescheduled, err := c.ensureScheduled(ctx, ns)
 	if err != nil {
 		return err
 	}
-
 	ns, err = c.ensureScheduledStatus(ctx, ns)
 	if err != nil {
 		return err
 	}
 
-	return c.enqueueResourcesForNamespace(ns)
+	// schedule resources in the namespace for rescheduling only if the namespace
+	// has not gotten rescheduling just now. We know that this namespace is requeued.
+	// Then we get another chance to reschedule the resources, and we avoid that
+	// resources are scheduled to a location in the stale namespace lister.
+	if !rescheduled {
+		return c.enqueueResourcesForNamespace(ns)
+	}
+
+	return nil
 }
 
 // enqueueResourcesForNamespace adds the resources contained by the given
-// namespace to the queue if the contents have never been successfully enqueued
-// or if the namespace's scheduling has changed since the last successful
-// enqueue. This ensures that the contents of a namespace eventually have the
-// same scheduling as their containing namespace.
+// namespace to the queue if there scheduling label differs from the namespace's.
 func (c *Controller) enqueueResourcesForNamespace(ns *corev1.Namespace) error {
-	lastScheduling, previouslyEnqueued := c.namepaceContentsEnqueuedFor(ns)
-	if previouslyEnqueued && lastScheduling == ns.Labels[ClusterLabel] {
-		klog.Infof("ANDY enqueueResourcesForNamespace short-circuiting %s|%s", logicalcluster.From(ns), ns.Name)
-		return nil
-	}
+	nsLocation := ns.Labels[ClusterLabel]
 
-	// TODO(marun) Consider only enqueueing items here if they're not already enqueued.
 	listers, notSynced := c.ddsif.Listers()
 	for gvr, lister := range listers {
-		klog.Infof("Enqueuing resources in namespace %s|%s for GVR %q", ns.ClusterName, ns.Name, gvr)
 		objs, err := lister.ByNamespace(ns.Name).List(labels.Everything())
 		if err != nil {
 			return err
 		}
+
+		first := true
 		for _, obj := range objs {
-			c.enqueueResource(gvr, obj)
+			u := obj.(*unstructured.Unstructured)
+			objLocation := u.GetLabels()[ClusterLabel]
+			if objLocation != nsLocation {
+				c.enqueueResource(gvr, obj)
+
+				if first {
+					klog.V(2).Infof("Enqueuing some %s in namespace %s|%s to schedule to %q", gvr, ns.ClusterName, ns.Name, nsLocation)
```

Comment:
I'm confused about this... we have obj - why log this the first time through at v(2)?

Reply:
Also it's only enqueuing this specific obj, not the entire gvr

Reply:
to see progress, but without too much noise with many objects.

Reply:
not perfect, but best I could come up with without logging per-object. Any better idea? We could log a limit number of resources, e.g. 10, in some slice in that line.

Reply:
I see - that's fine to do. Maybe add a comment indicating that?